### PR TITLE
Block tombstone GC if older data still exists in commitlog 

### LINKED ISSUE #14870: Writes in commitlog segments, that are pinned by another table, can be resurrected
Commitlog segments are only discarded when all tables have released their respective replay handles, that point to said segment. It is possible that one table, that receives very little writes, have very long lived memtables, that pin commitlog segments, containing writes to other tables. It is possible that these writes are covered with a tombstone, written in a later write. This tombstone can get to disk, be compacted and purged, while the commitlog segment containing the earlier (covered) writes is still pinned. A later unclean restart, can cause these segments to be replayed, resurrecting the data.

### PR DESCRIPTION
**Pull request by elcallio**
Fixes #14870 (alternative solution)

Blocks GC for compactions of tables where the sstable set holds data
that is potentially newer than data still held in non-deleted commitlog
segments. This prevents resurrection of data in the case of crash+replay
in cases where a very early compation removes tombstones, but CL data
is kept alive by other CF:s

Test: https://github.com/scylladb/scylla-dtest/pull/3352

Note: this messes with other tests assumptions on nodetool flush + compaction
and the result thereof, which is why these paths now also try to replace active 
segment in commitlog (if clean) to ensure we can in fact, as often as possible,
do compaction with GC and get the results we hope for.

### CODE REVIEW

File: replica/table.cc
Comment (denesb) [inline replica/table.cc:2824]: return false

File: replica/table.cc
Comment (denesb) [inline replica/table.cc:2843]: I think the two messages should be merged, perhaps use `\n` to separate the two parts.

File: replica/table.cc
Comment (avikivity) [inline replica/table.cc:2824]: false

File: replica/table.cc
Comment (denesb) [inline replica/table.cc:2853]: I think this might be too coarse. We should only check the actually compacted stables. Which might require this to be moved to `compaction.cc`.

**Review by denesb — COMMENTED** [review]
I like this direction more than the alternative (https://github.com/scylladb/scylladb/pull/14903).

File: replica/table.cc
Comment (elcallio) [reply replica/table.cc]: fixed.

Comment (elcallio) [reply replica/table.cc]: fixed

File: replica/table.cc
Comment (elcallio) [reply replica/table.cc:2843]: I really dislike multi-line log messages. Two messages looks nicer, but if you get a second vote I will change. :-)

File: replica/table.cc
Comment (elcallio) [reply replica/table.cc:2853]: Or add an interface method to process them, or add the set as a param to this call. In compaction manager we've lost a bunch of the info required here, which is a bit of a nuisance... (database, actual table, CL etc..)

Comment (elcallio) [reply replica/table.cc:2853]: Next problem is that this is checked from a bunch of places with various levels of info/object available. 
Maybe a compromise is to at least filter on eligible_for_compaction(sstable) here. Not sure how much/little this would filter out.

Comment (denesb) [reply replica/table.cc:2853]: We could add `std::vector<shared_sstable>` parameter. If the list is empty, we check the main+maint sets, as you do now. If the list is non-empty, we check only passed-in sstables.
Callers pass sstables if they have at hand, otherwise pass empty list.

Comment (elcallio) [reply replica/table.cc:2857]: I added a patch that extends the interface and hopefully sends the correct sstables as parameters when applicable. 
Tests still pass, so there is always that. :-)

File: conf/scylla.yaml
Comment (raphaelsc) [inline conf/scylla.yaml:603]: unrelated?

File: compaction/compaction.cc
Comment (raphaelsc) [inline compaction/compaction.cc:149]: you're interested in compacting set.

selector yields non-compacting sstables too

File: conf/scylla.yaml
Comment (elcallio) [reply conf/scylla.yaml:603]: woops. dtest.

File: compaction/compaction.cc
Comment (elcallio) [reply compaction/compaction.cc]: Fixed.

File: replica/table.cc
Comment (avikivity) [inline replica/table.cc:2847]: This looks expensive, and my intuition says it will usually return false (for many tables/tablets), no proof yet.

**Comment by elcallio** [comment]
Unfortunately, even with the extra CL segment rotation etc some compaction tests in dtest fail (esp. test_compaction_delete_tombstone_gc[immediate-SizeTieredCompactionStrategy]). 
Typically, we end up with things like various system tables blocking purge of user data.

File: replica/table.cc
Comment (elcallio) [reply replica/table.cc:2847]: The lookup is not very expensive, though obviously o(nsstables + nsegments). As for ratio between true/false, not sure. If compaction only selects older sstables for GC:ing compaction, it should be pretty safe (for real world cases), but obviously (as stated elsewhere in this PR), for tests which create artificial compactions early etc we end up with lock-outs quite a lot. 

Which illustrates the annoying balance here I guess... (Note, the truncation-using solution does not suffer from this issue, but generates extra writes and flushes - even if those could potentially be alleviated at least partially).

File: replica/table.cc
Comment (elcallio) [reply replica/table.cc:2857]: CL tests that is. Not all compaction tests. But those fail regardless with this approach...

**Comment by avikivity** [comment]
> Unfortunately, even with the extra CL segment rotation etc some compaction tests in dtest fail (esp. test_compaction_delete_tombstone_gc[immediate-SizeTieredCompactionStrategy]). Typically, we end up with things like various system tables blocking purge of user data.

Shouldn't a drain fix it?

**Comment by elcallio** [comment]
In truth, even the nodetool flush should handle it (since it is a flush-all), but it does not. Not sure yet as to why not. I hope it is a snafu on my end.

**Comment by avikivity** [comment]
Maybe it's a bug with flush not fully clearing commitlog. We should have a test that ensures it.

Another possibility is that during the flush we have some internal writes that go to commitlog; those internal writes are post-flush and so don't get flushed themselves, and tie up a commitlog segment. A candidate is the compaction history table (we can consider moving it to the schema commitlog if it's the culprit).

**Comment by avikivity** [comment]
Reasonable.

**Comment by elcallio** [comment]
Added a test fix in https://github.com/scylladb/scylla-dtest/pull/3451.

This somewhat shows that new data from latest memtable flush being involved in the compaction set causes us to fail compacting tombstones. Not sure how/if we can guarantee that at least some compactions (at some point) will restrict themselves to only old sstables.

File: replica/table.cc
Comment (raphaelsc) [reply replica/table.cc:2847]: Now I understand the fix better. 

We call tombstone_gc_enabled() for every partition containing an expired tombstone.

So the complexity is

O(partitions_with_tombstone * (nsstables + nsegments))

I am worried about the overhead for tombstone heavy workloads (can be default TTL, not only regular deletion), which is common. And those with small partitions will suffer even more.

we can probably mitigate it by doing commitlog check only once per compaction. worse granularity, but less overhead.

@elcallio why don't we use replay position stored in sstables to avoid replaying older data?

Comment (elcallio) [reply replica/table.cc:2847]: That is a good question. While the sstable only stores high RP (taken from memtable flushed), technically we could maybe use it the same way as her, but in inverse: I.e. if a higher RP that being replayed is in any live sstable, we can ignore the replay. Might be problematic however if a sstable is damaged/removed and we somehow rely on a combo of replay + repair to fix things up... Maybe that is never an issue...
I am probably forgetting something here...

**Comment by raphaelsc** [comment]
> Added a test fix in [scylladb/scylla-dtest#3451](https://github.com/scylladb/scylla-dtest/pull/3451).
> 
> This somewhat shows that new data from latest memtable flush being involved in the compaction set causes us to fail compacting tombstones. Not sure how/if we can guarantee that at least some compactions (at some point) will restrict themselves to only old sstables.

That's less likely to happen in practice as the lifetime of segment is usually lower than gc period. It's important that we make it visible that tombstone GC is being blocked by commitlog, a metric or warning in log. Will help the poor guy debugging it.

**Comment by denesb** [comment]
> > Added a test fix in [scylladb/scylla-dtest#3451](https://github.com/scylladb/scylla-dtest/pull/3451).
> > This somewhat shows that new data from latest memtable flush being involved in the compaction set causes us to fail compacting tombstones. Not sure how/if we can guarantee that at least some compactions (at some point) will restrict themselves to only old sstables.
> 
> That's less likely to happen in practice as the lifetime of segment is usually lower than gc period. It's important that we make it visible that tombstone GC is being blocked by commitlog, a metric or warning in log. Will help the poor guy debugging it.

We should also provide a means to unblock GC. Currently the only way is a clean restart. We already have `nodetool flush` but it allows flushing only a single table. We need a node-wide flush of all memtables.

File: replica/table.cc
Comment (denesb) [inline replica/table.cc:2852]: nit: I think `std::for_each()` (actually `std::ranges::for_each()`) would be a better fit here.

**Review by denesb — APPROVED** [review]
Not familiar with CL internals, but the change makes sense to me.

File: db/commitlog/commitlog.cc
Comment (avikivity) [inline db/commitlog/commitlog.cc:1936]: Q: let's say we're forcing a flush due to the hitting the hard limit. Will we deadlock? Or will just new commits wait until the flush completes (it's sad since it's a long wait, but okay since it's an edge case).

The fact that commitlog depends on activities that need commitlog is sad. Maybe we should direct all those internal table writes to the schema commitlog.

File: replica/table.cc
Comment (avikivity) [reply replica/table.cc:2847]: I believe the replay positions are useless. They tell us what commitlog data this sstable contains, but it doesn't allows us to exclude all commits before that position, because sstables contain filtered commitlog data. It's filtered by table (and that's something we know because we have the sstable->table association), but it can be filtered by token range (for compaction groups), but timestamp (for TWCS), and probably for other things. So to actually filter commits during replay, we need to know all the conditions that were used to create the sstable.

File: db/commitlog/commitlog.cc
Comment (elcallio) [reply db/commitlog/commitlog.cc:1936]: This will not, no, since the salient point is a segment exists and it has data. It might cause some really inefficient processing though, with writes + flushes of very small segments.

File: replica/table.cc
Comment (elcallio) [reply replica/table.cc:2847]: Actually, as long as sstables always retain the highest RP contained, it should be fine, since commit log is guarding against loosing data not written to sstable, not data potentially lost due to sstable being lost (a latter matter). A higher RP in any sstable always means that all lower RP:s have been written to disk (remember, we do not allow reordering when writing to memtable) at some point, in some sstable, thus data in CL should be safe to skip one way or another. 

In the truncation case, obviously this data is useless since sstables are gone, hence the truncation table. I still have a nagging memory that there is some reason we don't use sstable positions, that theý are not quite reliable, but I can't remember if/why... There is also the issue again of resharding, which makes them harder to compare...

File: db/commitlog/commitlog.cc
Comment (avikivity) [reply db/commitlog/commitlog.cc:1936]: I'm talking about this dependency loop:
1. commitlog is at the hard limit, so it wants to flush
2. flush now needs to allocate a new segment
3. we can't allocate a new segment because we're at the hard limit

Comment (elcallio) [reply db/commitlog/commitlog.cc:1936]: We don't actually allocate a new segment here. We just discard any existing one. Allocation will not happen until a new mutation comes in, which might block, yes, but this memtable flush/compaction will still be able to finish and release segments.

**Pull request closed without merging**
